api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -1,39 +1,218 @@
 package messagebuffer
 
 import (
-	"encoding/hex"
+	"fmt"
 	"sync"
+	"time"
 
+	"github.com/nknorg/nkn/v2/config"
 	"github.com/nknorg/nkn/v2/pb"
 )
 
-// MessageBuffer is the buffer to hold message for clients not online
-type MessageBuffer struct {
-	sync.Mutex
-	buffer map[string][]*pb.Relay
+type Item struct {
+	relay        *pb.Relay
+	creationTime time.Time
+	expiryTime   time.Time
 }
 
-// NewMessageBuffer creates a MessageBuffer
-func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+type queue struct {
+	parent *Buffer
+	item   *Item
+}
+
+type Buffer struct {
+	mux   sync.Mutex
+	queue map[string][]*queue
+	size  int
+}
+
+func New() *Buffer {
+	return &Buffer{
+		queue: make(map[string][]*queue, 0),
+	}
+}
+
+func getParent(index int) int {
+	if index <= 0 {
+		return -1
+	}
+
+	return (index - 1) / 2
+}
+
+func getChildren(index int) (int, int) {
+	tree := index * 2
+
+	return tree + 1, tree + 2
+}
+
+func (q *queue) isAfter(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return false
+	}
+
+	return true
+}
+
+func (q *queue) isAfterOrEqual(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return true
+	}
+
+	return true
+}
+
+func (m *Buffer) Purge() {
```

Comment:
Purge has O(N) time complexity, which is too high as a regular cleaning job

Reply:
correct. I have my trouble with purging this way. I am thinking about a better way to solve this. I left `Purge()` in there to demonstrate/test that an iteration over the whole map is needed to keep a clean cache state.

another way would be an irregular O(n) by doing the cleanup periodically and sacrificing accuracy.
yet another way would be to purge only when cache is full.

any ideas are welcome.

Reply:
You can use another heap (sorted by expiration time) for fast (O(log N)) purge. The caveat is that when removing an element from one heap, you need to remove it from the other one as well.